bridges/snowbridge/parachain/primitives/router/src/outbound/mod.rs

```diff
@@ -0,0 +1,283 @@
+
+pub struct EthereumBlobExporter<
+	UniversalLocation,
+	EthereumNetwork,
+	OutboundQueue,
+	AgentHashedDescription,
+>(PhantomData<(UniversalLocation, EthereumNetwork, OutboundQueue, AgentHashedDescription)>);
+
+impl<UniversalLocation, EthereumNetwork, OutboundQueue, AgentHashedDescription> ExportXcm
+	for EthereumBlobExporter<UniversalLocation, EthereumNetwork, OutboundQueue, AgentHashedDescription>
+where
+	UniversalLocation: Get<InteriorMultiLocation>,
+	EthereumNetwork: Get<NetworkId>,
+	OutboundQueue: SendMessage<Balance = u128>,
+	AgentHashedDescription: ConvertLocation<H256>,
+{
+	type Ticket = (Vec<u8>, XcmHash);
+
+	fn validate(
+		network: NetworkId,
+		_channel: u32,
+		universal_source: &mut Option<InteriorMultiLocation>,
+		destination: &mut Option<InteriorMultiLocation>,
+		message: &mut Option<Xcm<()>>,
+	) -> SendResult<Self::Ticket> {
+		let expected_network = EthereumNetwork::get();
+		let universal_location = UniversalLocation::get();
+
+		if network != expected_network {
+			log::trace!(target: "xcm::ethereum_blob_exporter", "skipped due to unmatched bridge network {network:?}.");
+			return Err(SendError::NotApplicable)
+		}
+
+		let dest = destination.take().ok_or(SendError::MissingArgument)?;
+		if dest != Here {
+			log::trace!(target: "xcm::ethereum_blob_exporter", "skipped due to unmatched remote destination {dest:?}.");
+			return Err(SendError::NotApplicable)
+		}
+
+		let (local_net, local_sub) = universal_source
+			.take()
+			.ok_or_else(|| {
+				log::error!(target: "xcm::ethereum_blob_exporter", "universal source not provided.");
+				SendError::MissingArgument
+			})?
+			.split_global()
+			.map_err(|()| {
+				log::error!(target: "xcm::ethereum_blob_exporter", "could not get global consensus from universal source '{universal_source:?}'.");
+				SendError::Unroutable
+			})?;
+
+		if Ok(local_net) != universal_location.global_consensus() {
+			log::trace!(target: "xcm::ethereum_blob_exporter", "skipped due to unmatched relay network {local_net:?}.");
+			return Err(SendError::NotApplicable)
+		}
+
+		let para_id = match local_sub {
+			X1(Parachain(para_id)) => para_id,
+			_ => {
+				log::error!(target: "xcm::ethereum_blob_exporter", "could not get parachain id from universal source '{local_sub:?}'.");
+				return Err(SendError::MissingArgument)
+			},
+		};
+
+		let message = message.take().ok_or_else(|| {
+			log::error!(target: "xcm::ethereum_blob_exporter", "xcm message not provided.");
+			SendError::MissingArgument
+		})?;
+
+		let mut converter = XcmConverter::new(&message, &expected_network);
+		let (agent_execute_command, message_id) = converter.convert().map_err(|err|{
+			log::error!(target: "xcm::ethereum_blob_exporter", "unroutable due to pattern matching error '{err:?}'.");
+			SendError::Unroutable
+		})?;
+
+		// local_sub is relative to the relaychain. No conversion needed.
+		let local_sub_location: MultiLocation = local_sub.into();
```

Comment:
Could you, please, check me here. Looking at the `snowbridge-system` pallet, I see that to compute `agent_id` we call the `AgentIdOf::convert_location(MultiLocation { parents: 1, interior: X1(Parachain(<para-id>)) })`. But here we do `split_global()` to get the `local_sub`, so it is the `InteriorMultiLocation`. And then we use the `MultiLocation::from(local_sub)` (this line), which constructs `MultiLocation` with parents = 0. Shouldn't it be the same `agent_id` everywhere? Om am I missing something here?

Reply:
Yes, it's a bit confusing here. Actually, for asset_hub we hardcode the `agent_id` derived from  `AgentIdOf::convert_location(MultiLocation { parents: 0, interior: X1(Parachain(1000)) })`(i.e. in perspective of relay chain, not the sibling) and with that to initialize the gateway contract. 

We may need to deal with it consistently.

Reply:
@svyatonik It's confirmed and addressed in https://github.com/Snowfork/snowbridge/pull/1077.
Thanks for all your comments, insightful and helpful!